compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2108,11 +2108,20 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 if let Some((expected, found)) = trace.values.ty() {
                     match (expected.kind(), found.kind()) {
                         (ty::Tuple(_), ty::Tuple(_)) => {}
-                        // If a tuple of length one was expected and the found expression has
-                        // parentheses around it, perhaps the user meant to write `(expr,)` to
-                        // build a tuple (issue #86100)
                         (ty::Tuple(fields), _) => {
-                            self.emit_tuple_wrap_err(&mut err, span, found, fields)
+                            if fields.len() == 0 {
+                                err.span_suggestion(
```

Comment:
Check if the `trace`'s obligation cause code is `BlockTailExpression`. If so, then we should be able to [get the expr from the `HirId`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/traits/enum.ObligationCauseCode.html#variant.BlockTailExpression).

We should double check that the expr's type is equal to `expected` (use [this method](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TypeckResults.html#method.node_type) -- you can get `TypeckResults` from infer ctxt). 

Then get that expr's span so we can suggest `;`.